substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -1417,9 +1460,25 @@ where
 		// transactions with those hashes from the retracted blocks.
 		let mut pruned_log = HashSet::<ExtrinsicHash<ChainApi>>::new();
 
+		// Create a map from enacted blocks' extrinsics to their `provides`
+		// tags based on inactive views.
+		let known_provides_tags = self.collect_provides_tags_from_view_store(tree_route).await;
+		debug!(target: LOG_TARGET, "update_view_with_fork: txs to tags map length: {}", known_provides_tags.len());
 		future::join_all(tree_route.enacted().iter().map(|hn| {
 			let api = api.clone();
-			async move { (hn, crate::prune_known_txs_for_block(hn, &*api, &view.pool).await) }
+			let known_provides_tags = known_provides_tags.clone();
+			async move {
+				(
+					hn,
+					crate::prune_known_txs_for_block(
```

Comment:
Nit, not sure if worthy refactoring:

`prune_known_txs_for_block` function also fetches block bodies from enacted fork here:
https://github.com/paritytech/polkadot-sdk/blob/7e5b993c32565f753e7904074ed3e5775a2838d5/substrate/client/transaction-pool/src/single_state_txpool/single_state_txpool.rs#L606-L613

Technically we could re-used the transactions fetched from enacted fork which is already done in `collect_provides_tags_from_view_store` above to avoid duplicated call to block_body. (`HashMap<BlockHash, Vec<Transaction>>` could be returned.)

This probably goes into the cache - and the second call will fetch from there - so the final cost would not be that big.

Would be good to have second opinion (cc: @skunert).

Reply:
I don't think we cache block bodies in the backend. Maybe some rocksdb or parityDB cache takes effect. But in our substrate backend, we only cache headers and pinned bodies. But we can not assume a pinned status for the bodies here. So optimization sounds useful.